portfolio/src/main/java/com/google/sps/data/ReviewDaoDatastore.java

```diff
@@ -0,0 +1,205 @@
+package com.google.sps.data;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.Filter;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.ImmutableSet.Builder;
+import com.google.sps.data.Book;
+import com.google.sps.data.Review;
+import java.lang.Exception;
+import java.lang.NumberFormatException;
+import java.util.Set;
+
+public class ReviewDaoDatastore implements ReviewDao {
+  private DatastoreService datastore;
+  private static final String ENTITY_KIND = "Review";
+  private static final String ISBN_PROPERTY = "ISBN";
+  private static final String FULLTEXT_PROPERTY = "FullText";
+  private static final String USEREMAIL_PROPERTY = "UserEmail";
+  private static final String DEFAULT_EMAIL = "unknown@email.com";
+
+  public ReviewDaoDatastore() {
+    datastore = DatastoreServiceFactory.getDatastoreService();
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void uploadAll(Book book) {
+    int userNumber = 0;
+    for (String reviewStr : book.reviews()) {
+      Review review = Review.builder()
+                          .fullText(reviewStr)
+                          .isbn(book.isbn())
+                          .email(DEFAULT_EMAIL + Integer.toString(userNumber))
+                          .build();
+      datastore.put(reviewToEntity(review));
+      userNumber++;
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void uploadNew(Review review) throws Exception {
+    if (reviewExists(review.isbn(), review.email())) {
+      throw new Exception(
+          "Review exists for ISBN:" + review.isbn() + ", by user " + review.email());
+    }
+    datastore.put(reviewToEntity(review));
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void updateReview(Review review) throws EntityNotFoundException {
+    String isbn = review.isbn();
+    String email = review.email();
+    String fullText = review.fullText();
+
+    if (!reviewExists(isbn, email)) {
+      throw new EntityNotFoundException(createKey(isbn, email));
+    }
+    datastore.delete(createKey(isbn, email));
```

Comment:
is it necessary to delete an entity in order to update it? i think you might just be able to do a ,put
https://cloud.google.com/appengine/docs/standard/java/datastore/creating-entities#Updating_entities